### PR TITLE
Fix flaky test test_concurrent_snapshot_packaging 

### LINKED ISSUE #14945: Failing test_concurrent_snapshot_packaging
#### Problem

Test failed in automation test_concurrent_snapshot_packaging

```
[2021-01-29T17:31:34.250967205Z ERROR solana_runtime::append_vec] AppendVec failed to remove "/tmp/.tmp7EsPrm/488.491": Os { code: 2, kind: NotFound, message: "No such file or directory" }
--
  | [2021-01-29T17:31:34.251118130Z ERROR solana_runtime::append_vec] AppendVec failed to remove "/tmp/.tmp7EsPrm/4828.4831": Os { code: 2, kind: NotFound, message: "No such file or directory" }
  | [2021-01-29T17:31:34.251238547Z ERROR solana_runtime::append_vec] AppendVec failed to remove "/tmp/.tmp7EsPrm/5324.5327": Os { code: 2, kind: NotFound, message: "No such file or directory" }
  | [2021-01-29T17:31:34.251351109Z ERROR solana_runtime::append_vec] AppendVec failed to remove "/tmp/.tmp7EsPrm/287.290": Os { code: 2, kind: NotFound, message: "No such file or directory" }
  | [2021-01-29T17:31:34.251474511Z ERROR solana_runtime::append_vec] AppendVec failed to remove "/tmp/.tmp7EsPrm/552.555": Os { code: 2, kind: NotFound, message: "No such file or directory" }
  | test tests::V1_2_0_Devnet::test_slots_to_snapshot ... ok
  |  
  | failures:
  |  
  | failures:
  | tests::V1_2_0_Development::test_concurrent_snapshot_packaging
```

https://buildkite.com/solana-labs/solana/builds/39112#dac81271-df67-4c99-95fb-795abbaeca94

#### Proposed Solution

Debug and fix

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
`test_concurrent_snapshot_packaging` is flaky due to occasional failure here: https://github.com/solana-labs/solana/blob/master/runtime/src/snapshot_utils.rs#L867. This is because `get_snapshot_storages()` filters out some storage files that contain no accounts https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L4623, so the saved directory here https://github.com/solana-labs/solana/blob/e59a24d9f9925c4fe7ca70cdbc732e4a967475a1/core/tests/snapshots.rs#L374 copies extra files that won't show up in the snapshot, causing the mismatch.

#### Summary of Changes
Fix test to only save off files returned by `get_snapshot_storages`

Fixes #14945

### CODE REVIEW

File: core/tests/snapshots.rs [hunk 1]
```diff
@@ -370,8 +369,24 @@ mod tests {
...
+                let snapshot_storage_files: HashSet<_> = bank_forks[slot]
+                    .get_snapshot_storages()
+                    .into_iter()
+                    .flatten()
+                    .map(|s| s.get_path())
+                    .collect();
+
+                // Only save off the files returned by `get_snapshot_storages`. This is because
+                // some of the storage entries in the accounts directory may be filtered out by
+                // `get_snapshot_storages()` and will not be included in the snapshot.
```
Comment (ryoqun) [inline core/tests/snapshots.rs:382]: nits: Add this? `// So, copying naitvely everything in accounts_dir could lead to test failure by mismatch`

File: core/tests/snapshots.rs [hunk 2]
```diff
@@ -480,9 +496,7 @@ mod tests {
         snapshot_utils::verify_snapshot_archive(
             saved_archive_path.unwrap(),
             saved_snapshots_dir.path(),
-            saved_accounts_dir
-                .path()
-                .join(accounts_dir.path().file_name().unwrap()),
```
Comment (ryoqun) [inline core/tests/snapshots.rs:485]: so, this `.join` isn't needed to begin with? ;)

Comment (ryoqun) [reply core/tests/snapshots.rs:485]: oh, I see now. it looks like new code does 1-level shallower copies.

**Review by ryoqun — APPROVED** [review]
LGTM!

Thanks so much for debugging this. I couldn't make time for this for days...

**Comment by ryoqun** [comment]
(come to think of it, maybe this test code problem got more susceptible because of now always-on AccountCaching by default?)

**Comment by carllin** [comment]
> (come to think of it, maybe this test code problem got more susceptible because of now always-on AccountCaching by default? hint: random flushing or recycled AppendVec?)

Hehe account flushing is actually still off by default in AccountsDb::default(), it's just enabled by default in the validator config

**Pull request merged**